core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/signer/AwsChunkedV4aPayloadSigner.java

```diff
@@ -93,19 +111,98 @@ private void setupSigExt(ChunkedEncodedInputStream.Builder builder, RollingSigne
+            () -> trailers.stream().map(TrailerProvider::get).collect(Collectors.toMap(Pair::left, Pair::right));
+        Supplier<byte[]> sigSupplier = () -> rollingSigner.sign(headersSupplier.get());
 
         builder.addTrailer(
             () -> Pair.of(
                 "x-amz-trailer-signature",
-                Collections.singletonList(new String(rollingSigner.sign(trailers), StandardCharsets.UTF_8))
+                Collections.singletonList(new String(sigSupplier.get(), StandardCharsets.UTF_8))
             )
         );
     }
 
-    private void setupChecksumTrailer(ChunkedEncodedInputStream.Builder builder) {
-        // TODO(sra-identity-and-auth): Set up checksumming of chunks and add as a trailer
+    /**
+     * Add the checksum as a chunk-trailer and add it to the request's trailer header.
+     * <p>
+     * The checksum-algorithm MUST be set if this is called, otherwise it will throw.
+     */
+    private void setupChecksumTrailerIfNeeded(ChunkedEncodedInputStream.Builder builder, SdkHttpRequest.Builder request) {
+        if (checksumAlgorithm == null) {
+            return;
+        }
+        SdkChecksum sdkChecksum = fromChecksumAlgorithm(checksumAlgorithm);
+        ChecksumInputStream checksumInputStream = new ChecksumInputStream(
+            builder.inputStream(),
+            Collections.singleton(sdkChecksum)
+        );
+        String checksumHeaderName = checksumHeaderName(checksumAlgorithm);
+
+        TrailerProvider checksumTrailer = () -> Pair.of(
+            checksumHeaderName,
+            Collections.singletonList(sdkChecksum.getChecksum())
+        );
+
+        request.appendHeader(X_AMZ_TRAILER, checksumHeaderName);
+        builder.inputStream(checksumInputStream).addTrailer(checksumTrailer);
+    }
+
+    /**
+     * Create chunk-trailers for each pre-existing trailer given in the request.
+     * <p>
+     * However, we need to validate that these are valid trailers. Since aws-chunked encoding adds the checksum as a trailer, it
+     * isn't part of the request headers, but other trailers MUST be present in the request-headers.
+     */
+    private void setupPreExistingTrailers(ChunkedEncodedInputStream.Builder builder, SdkHttpRequest.Builder request) {
+        List<String> trailerHeaders = request.matchingHeaders(X_AMZ_TRAILER);
+
+        for (String header : trailerHeaders) {
+            List<String> values = request.matchingHeaders(header);
+            if (values.isEmpty()) {
+                throw new IllegalArgumentException(header + " must be present in the request headers to be a valid trailer.");
+            }
+
+            // Add the trailer to the aws-chunked stream-builder, and remove it from the request headers
+            builder.addTrailer(() -> Pair.of(header, values));
+            request.removeHeader(header);
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    static class Builder {
```

Comment:
Can be final. Guess it's `package-private` because it's internal and won't be accessed outside of this package?